gateway/src/apicast/policy/rate_limit/rate_limit.lua

```diff
@@ -0,0 +1,222 @@
+      return true, nil
+    end,
+    get = function(_, key)
+      local val = redis:get(key)
+      if type(val) == "userdata" then
+        return nil
+      end
+      return val
+    end
+  }
+end
+
+local function error(logging_only, status_code)
+  if not logging_only then
+    return ngx.exit(status_code)
+  end
+end
+
+function _M.new(config)
+  local self = new()
+  self.config = config or {}
+  self.limiters = config.limiters
+  self.redis_url = config.redis_url
+  self.status_code_rejected = config.status_code_rejected or 429
+  self.logging_only = config.logging_only or false
+
+  return self
+end
+
+function _M:access()
+  local limiters = {}
+  local keys = {}
+
+  if not self.redis_url then
+    -- Only one (the first) limiter is enable.
+    -- Key will be shdict_key ('limiter').
+    local lim, initerr = init_limiter(self.limiters[1])
+    if not lim then
+      ngx.log(ngx.ERR, "unknown limiter: ", initerr)
+      error(self.logging_only, 500)
+      goto done
+    end
+
+    limiters[1] = lim
+    keys[1] = shdict_key
+
+  else
+    for _, limiter in ipairs(self.limiters) do
+      local lim, initerr = init_limiter(limiter)
+      if not lim then
+        ngx.log(ngx.ERR, "unknown limiter: ", limiter.name, ", err: ", initerr)
+        error(self.logging_only, 500)
+        goto done
+      end
+
+      local rediserr
+      lim.dict, rediserr = redis_shdict(self.redis_url)
+      if not lim.dict then
+        ngx.log(ngx.ERR, "failed to connect Redis: ", rediserr)
+        error(self.logging_only, 500)
+        goto done
+      end
+
+      limiters[#limiters + 1] = lim
+      keys[#keys + 1] = limiter.key
+
+    end
+  end
+
+  local states = {}
+  local connections_committed = {}
+  local keys_committed = {}
+
+  local delay, comerr = limit_traffic.combine(limiters, keys, states)
+  if not delay then
+    if comerr == "rejected" then
+      ngx.log(ngx.WARN, "Requests over the limit.")
+      error(self.logging_only, self.status_code_rejected)
+      goto done
+    end
+    ngx.log(ngx.ERR, "failed to limit traffic: ", comerr)
+    error(self.logging_only, 500)
+    goto done
+  end
+
+  for i, lim in ipairs(limiters) do
+    if lim.is_committed and lim:is_committed() then
+      connections_committed[#connections_committed + 1] = lim
+      keys_committed[#keys_committed + 1] = keys[i]
+    end
+  end
+
+  if next(connections_committed) ~= nil then
+    local ctx = ngx.ctx
+    ctx.limiters = connections_committed
+    ctx.keys = keys_committed
+  end
+
+  if delay >= 0.001 then
+    ngx.log(ngx.WARN, 'need to delay by: ', delay, 's, states: ', table.concat(states, ", "))
+    ngx.sleep(delay)
+  end
+
+  ::done::
+end
+
+local function checkin(_, ctx, time, semaphore, redis_url, logging_only)
+  local limiters = ctx.limiters
+  local keys = ctx.keys
+  local latency = tonumber(time)
+
+  for i, lim in ipairs(limiters) do
+    if redis_url then
+      local rediserr
+      lim.dict, rediserr = redis_shdict(redis_url)
```

Comment:
This is needed because the connection has been already closed.

Reply:
This initializes a Redis client instance on every request. It might be better to have a connection pool and fetch and return connections as needed, but maybe we can leave this for a future PR.

Reply:
@davidor I'd like to leave this for a future PR because it's a little bit difficult.